### ISSUE TITLE
The vm does not throw when `fromEnvironment`/`hasEnvironment` is invoked at run time

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#41168]
There is a discrepancy between the actual and the documented behavior in the declaration of the special constructors `{bool,int,String}.fromEnvironment` and `bool.hasEnvironment`, indicating that they must throw if invoked at run time. In other words, they are for constant expressions only, but the vm does not throw when they are invoked at run time.

This issue serves as a reference to give when approving tests that fail for this reason. It is not expected that the throwing behavior is implemented soon, or perhaps at all.

**Comment by srawlins** [comment]
@eernstg Should the analyzer be reporting an error as well for `new String.fromEnvironment("x")`? I see exactly one instance of `fromEnvironment` in the spec. In fact, there are only ~4 references in the language repo, and none of them define this behavior.

**Comment by eernstg** [comment]
No, I agree that there is no justification for a compile-time error from the analyzer here. The location where the language specification mentions `fromEnvironment` is commentary, so the whole `fromEnvironment`/`hasEnvironment` feature is a core library feature and not a language feature, and a compile-time error message must report a fault in the program which is defined to be an error in the language.

Another matter is that it would be helpful to emit a hint whenever it is detected that one of these constructors will be invoked at run time, because `dart2js` compiled programs _will_ throw, and the vm may also start doing that at some point.

**Comment by FMorschel** [comment]
A related Flutter issue with 100 upvotes:

- https://github.com/flutter/flutter/issues/55870

**Comment by FMorschel** [comment]
I do think a lint would be helpful here. It doesn't need to be an `analyzer` warning. But maybe a `core` lint (if we ever do something here)? 

CC @bwilkerson

**Comment by lrhn** [comment]
The VM has always allowed non-`const` invocation of the `fromEnvironment` constructors. At a time, so did the web compilers. It was a web compiler which first introduced the "must be `const`" requirement.

The ability to pass different values at runtime than the ones used while compiling, as reflected in constants, is what gives this issue.
If the VM remembered the compilation environment when it's accessed as non-`const`, there wouldn't be a problem, except that there might be scripts depending on the current behavior.

Using non-`const` invocations makes sense for the standalone VM, where there is no distinction between compile-time and runtime, so maybe the VM AOT compiler should reject non-`const` uses. Or the VM could deprecate the feature entirely, and require you to use `const`.

See also: https://github.com/dart-lang/sdk/issues/36482

**Comment by FMorschel** [comment]
I found this issue because I tried to evaluate a `String.fromEnvironment('FOO')` invocation while around a `const` code that used it but got _nothing_, and after stepping over with the debugger, the value was there. So this was a bit confusing.
